modules/core/src/main/java/org/apache/ignite/internal/client/thin/io/gridnioserver/GridNioServerClientConnectionMultiplexer.java

```diff
@@ -0,0 +1,177 @@
+        IgniteLogger gridLog = new JavaLogger(false);
+
+        ClientMessageDecoder decoder = new ClientMessageDecoder();
+
+        GridNioFilter[] filters;
+
+        GridNioFilter codecFilter = new GridNioCodecFilter(new GridNioParser() {
+            @Override
+            public @Nullable Object decode(GridNioSession ses, ByteBuffer buf) throws IOException, IgniteCheckedException {
+                byte[] bytes = decoder.apply(buf);
+
+                return ByteBuffer.wrap(bytes).order(ByteOrder.nativeOrder());
+            }
+
+            @Override
+            public ByteBuffer encode(GridNioSession ses, Object msg) throws IOException, IgniteCheckedException {
+                return (ByteBuffer)msg;
+            }
+        }, gridLog, false);
+
+//        if (sslCtx != null) {
+//            GridNioSslFilter sslFilter = new GridNioSslFilter(sslCtx, true, ByteOrder.nativeOrder(), gridLog);
+//
+//            sslFilter.directMode(false);
+//
+//            filters = new GridNioFilter[]{codecFilter, sslFilter};
+//        }
+//        else
+        filters = new GridNioFilter[]{codecFilter};
+
+        try {
+            srv = GridNioServer.<ByteBuffer>builder()
+                    .address(InetAddress.getLoopbackAddress()) // TODO: Remove?
+                    .port(CLIENT_MODE_PORT)
+                    .listener(new GridNioServerListener<ByteBuffer>() {
+                        @Override
+                        public void onConnected(GridNioSession ses) {
+                            System.out.println("Connect");
+                        }
+
+                        @Override
+                        public void onDisconnected(GridNioSession ses, @Nullable Exception e) {
+                            System.out.println("Disconnect");
+                        }
+
+                        @Override
+                        public void onMessageSent(GridNioSession ses, ByteBuffer msg) {
+
+                        }
+
+                        @Override
+                        public void onMessage(GridNioSession ses, ByteBuffer msg) {
+                            GridNioServerClientConnection conn = ses.meta(GridNioServerClientConnection.SES_META_CONN);
+
+                            conn.onMessage(msg);
+                        }
+
+                        @Override
+                        public void onSessionWriteTimeout(GridNioSession ses) {
+
+                        }
+
+                        @Override
+                        public void onSessionIdleTimeout(GridNioSession ses) {
+
+                        }
+
+                        @Override
+                        public void onFailure(FailureType failureType, Throwable failure) {
+                            System.out.println("Fail");
+                        }
+                    })
+                    .filters(filters)
+                    .logger(gridLog)
+                    .selectorCount(1) // TODO: Get from settings
```

Comment:
To be honest, there is no need to create more than one selector.

Reply:
@ivandasch I don't think so:
* With Partition Awareness a single `IgniteClient` connects to all server nodes
* `IgniteClient` can be used from multiple threads

Single worker thread can easily become a bottleneck.

Reply:
This is not a bottleneck at all, it's common practice to create 1 thread per selector and for clients 1 selector is more than enough. Remember, tarantool or redis use 1 thread for all network activity and this is more than enough. This is not even a server (sometimes it's desirable to split acceptor task to another selector)

Reply:
>> With Partition Awareness a single IgniteClient connects to all server nodes
1 thread easily can handle thousands connections. Look and nodejs for another example.

Reply:
Ok, you may be right. I'll do some benchmarks and we'll see.

> nodejs for another example

Not a good example because of single-threaded javascript nature; also, not anymore https://nodejs.org/api/worker_threads.html

Reply:
> Workers (threads) are useful for performing CPU-intensive JavaScript operations. They will not help much with I/O-intensive work. Node.js’s built-in asynchronous I/O operations are more efficient than Workers can be.

And, AFAIU you use ForkJoinPool for user code...

Reply:
Yes, GridNioServer workers are used solely for encoding and decoding requests in form of byte buffers and arrays - essentially copying some bytes around, which is trivial, so maybe one thread is really enough. We'll see. I'm happy to use a single worker and avoid one more config parameter.

Reply:
@ivandasch I've did some basic benchmarks (see JmhThinClientCacheBenchmark) and could not come up with a config where multiple selectors work faster than one (various cluster sizes, thread counts, entry sizes). You are right, let's keep this hardcoded to `1` and not bother with making it configurable.